packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -70,42 +75,243 @@ contract MixinExchangeCore is
+        if (order.takerAssetAmount == 0) {
+            orderInfo.orderStatus = uint8(Status.ORDER_INVALID_TAKER_ASSET_AMOUNT);
+            return orderInfo;
+        }
+
+        // Validate order expiration
+        if (block.timestamp >= order.expirationTimeSeconds) {
+            orderInfo.orderStatus = uint8(Status.ORDER_EXPIRED);
+            return orderInfo;
+        }
+
+        // Check if order has been cancelled
+        if (cancelled[orderInfo.orderHash]) {
+            orderInfo.orderStatus = uint8(Status.ORDER_CANCELLED);
+            return orderInfo;
+        }
+        if (makerEpoch[order.makerAddress] > order.salt) {
+            orderInfo.orderStatus = uint8(Status.ORDER_CANCELLED);
+            return orderInfo;
+        }
+
+        // Fetch filled amount and validate order availability
+        orderInfo.orderFilledAmount = filled[orderInfo.orderHash];
+        if (orderInfo.orderFilledAmount >= order.takerAssetAmount) {
+            orderInfo.orderStatus = uint8(Status.ORDER_FULLY_FILLED);
+            return orderInfo;
+        }
+
+        // All other statuses are ruled out: order is Fillable
+        orderInfo.orderStatus = uint8(Status.ORDER_FILLABLE);
+        return orderInfo;
+    }
+
+    /// @dev Calculates amounts filled and fees paid by maker and taker.
+    /// @param order to be filled.
+    /// @param orderStatus Status of order to be filled.
+    /// @param orderFilledAmount Amount of order already filled.
+    /// @param takerAssetFillAmount Desired amount of order to fill by taker.
+    /// @return status Return status of calculating fill amounts. Returns Status.SUCCESS on success.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function calculateFillResults(
+        Order memory order,
+        uint8 orderStatus,
+        uint256 orderFilledAmount,
+        uint256 takerAssetFillAmount
+    )
+        public
+        pure
+        returns (
+            uint8 status,
+            FillResults memory fillResults
+        )
+    {
+        // Fill amount must be greater than 0
+        if (takerAssetFillAmount == 0) {
+            status = uint8(Status.TAKER_ASSET_FILL_AMOUNT_TOO_LOW);
+            return;
+        }
+
+        // Ensure the order is fillable
+        if (orderStatus != uint8(Status.ORDER_FILLABLE)) {
+            status = orderStatus;
+            return;
+        }
+
+        // Compute takerAssetFilledAmount
+        uint256 remainingTakerAssetAmount = safeSub(order.takerAssetAmount, orderFilledAmount);
+        uint256 takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetAmount);
+
+        // Validate fill order rounding
+        if (isRoundingError(
+            takerAssetFilledAmount,
+            order.takerAssetAmount,
+            order.makerAssetAmount))
+        {
+            status = uint8(Status.ROUNDING_ERROR_TOO_LARGE);
+            return (status, fillResults);
```

Comment:
Why do we `return (status, fillResults)` here but just `return` in the previous `if` blocks? (I actually prefer this, so would change the previous 2).